### ISSUE TITLE
`udev` rule to `g+rw` the `/dev/uinput` does not work

### ISSUE THREAD

**Issue by glaeqen** [ReimuNotMoe/ydotool#210]
On Arch since very recently `ydotoold` started to fail to start because the udev rule for the  `/dev/uinput` file seized to take an effect. It's permissions remain to be `600` which means that the user with an `input` group cannot access it.

I think that I've got hit by [this](https://github.com/chrippa/ds4drv/issues/93#issuecomment-265300511) problem. Not sure what changed upstream that it started to happen just lately but here we are. Surprisingly, if  I `chown g+rw /dev/uinput` manually, `ydotoold` starts to work correctly even though I don't see the `uinput` module loaded at all. Also, just like the person from the other thread suggested, if I `modprobe uinput` manually, the permissions are applied correctly. Latter I can understand but former? How `/dev/uinput` can work without a kernel module being loaded to begin with? Curious.

I'm creating this issue to help people around trying to find a solution and have a discussion how to tackle this problem correctly. Is this something that people using other distros even witness? Do we need to create a file forcing a module load (like in a linked suggestion) or something else should be undertaken?

**Comment by yourealwaysbe** [comment]
I have the same issue on Arch Linux.

I tried to fix it by adding `uinput` to `/etc/modules-load.d/uinput.conf` but found that even though the module was loaded, the udev rule did not change the permissions. It seems like the module is loaded before the udev rules are handled, because if i then unload then reload uinput after logging in, the udev rule kicks in. The same happened if i used `/etc/modules-load.d/99-uninput.conf` to try to delay loading until last.

Different to your case, i found that running `chmod 660 /dev/uinput` and then starting `ydotoold` succeeds and uinput is loaded. Kind of like the module is being loaded lazily, but ydotoold detects early that it doesn't have permission and gives up before the load and udev permission change can kick in..

**Comment by yourealwaysbe** [comment]
Looking at [the code](https://github.com/ReimuNotMoe/ydotool/blob/a665655c7aeb837b3822d722ebe638198b382611/Daemon/ydotoold.c#L265C13-L270), there's no early failure because of the permissions on `/dev/uinput`. It just tries to open it and handles the failure when it occurs.

So i guess there is a cyclic issue: uinput won't load until something that has permission tries to open it, but permission won't be granted until uinput is loaded.

**Comment by yourealwaysbe** [comment]
For now i am adding to my sudoer file, where 1000:

    Cmnd_Alias YDOTOOLD_MYUSER = /sbin/ydotoold --socket-own 1000\:1000
    myuser ALL=(ALL) ... NOPASSWD:SETENV:YDOTOOLD_MYUSER

and starting

    $ sudo ydotool --socket-own 1000:1000

when my desktop starts (using exec in my Sway config).

(Where 1000 is my user id.)

**Comment by yourealwaysbe** [comment]
I created an [Arch Linux bug report](https://bugs.archlinux.org/task/79391) since the udev permissions change is part of the Linux package not ydotool afaik.

**Comment by glaeqen** [comment]
Can also confirm that downgrade to `systemd` version `253` helped. Thanks for help ❤️ I might try to bisect the systemd in spare time unless you already found the guilty commit?

**Comment by yourealwaysbe** [comment]
Just tested a `systemd-git` installation (required `pwquality`, `libselinux`, and `apparmor` not mentioned by the PKGBUILD) and it seems to be fixed there.

**Comment by leath-dub** [comment]
If anyone is wanting another temp fix, You do this to the ydotool binary
```bash
chown root `which ydotool`
chgrp root `which ydotool`
chmod u+s `which ydotool`
```
This just forces sid to be root. Ofc be aware that this gives ydotool root priviliges

**Comment by glaeqen** [comment]
> If anyone is wanting another temp fix, You do this to the ydotool binary
> 
> ```shell
> chown root `which ydotool`
> chgrp root `which ydotool`
> chmod u+s `which ydotool`
> ```
> 
> This just forces sid to be root. Ofc be aware that this gives ydotool root priviliges

That's an option although I'd rather suggest downgrading `systemd` to version `253` and waiting for the major release that supposedly has the regression fixed.

**Comment by leath-dub** [comment]
> > If anyone is wanting another temp fix, You do this to the ydotool binary
> > ```shell
> > chown root `which ydotool`
> > chgrp root `which ydotool`
> > chmod u+s `which ydotool`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This just forces sid to be root. Ofc be aware that this gives ydotool root priviliges
> 
> That's an option; the alternative workaround is downgrading `systemd` to version `253` and waiting for the major release that supposedly has the regression fixed.

I feel like its good to avoid downgrading at all costs, but yeah I did see that solution too - didn't want to have some new issues though.

**Comment by glaeqen** [comment]
Issue has been resolved since `systemd` version 254.2. Closing.
